packages/@aws-cdk/aws-kms/test/key.test.ts

```diff
@@ -582,6 +582,321 @@ describe('imported keys', () => {
+      key = kms.Key.fromCfnKey(cfnKey);
+    });
+
+    describe('and using it for grantDecrypt() on a Role', function () {
+      beforeEach(() => {
+        const role = new iam.Role(stack, 'Role', {
+          assumedBy: new iam.AnyPrincipal(),
+        });
+        key.grantDecrypt(role);
+      });
+
+      test('creates the correct IAM Policy', () => {
+        expectCdk(stack).to(haveResourceLike('AWS::IAM::Policy', {
+          PolicyDocument: {
+            Statement: [
+              {
+                Action: 'kms:Decrypt',
+                Effect: 'Allow',
+                Resource: {
+                  'Fn::GetAtt': ['CfnKey', 'Arn'],
+                },
+              },
+            ],
+          },
+        }));
+      });
+
+      test('correctly mutates the Policy of the underlying CfnKey', () => {
+        expectCdk(stack).to(haveResourceLike('AWS::KMS::Key', {
+          KeyPolicy: {
+            Statement: [
+              {
+                Action: 'kms:*',
+                Effect: 'Allow',
+                Principal: {
+                  AWS: {
+                    'Fn::Join': ['', [
+                      'arn:',
+                      { Ref: 'AWS::Partition' },
+                      ':iam::',
+                      { Ref: 'AWS::AccountId' },
+                      ':root',
+                    ]],
+                  },
+                },
+                Resource: '*',
+              },
+              {
+                Action: 'kms:Decrypt',
+                Effect: 'Allow',
+                Principal: {
+                  AWS: {
+                    'Fn::GetAtt': ['Role1ABCC5F0', 'Arn'],
+                  },
+                },
+                Resource: '*',
+              },
+            ],
+            Version: '2012-10-17',
+          },
+        }));
+      });
+    });
+  });
+
+  describe("called with a CfnKey that has an 'Fn::If' passed as the KeyPolicy", () => {
+    beforeEach(() => {
+      cfnKey = new kms.CfnKey(stack, 'CfnKey2', {
+        keyPolicy: cdk.Fn.conditionIf(
+          'AlwaysTrue',
+          {
+            Statement: [
+              {
+                Action: 'kms:action1',
+                Effect: 'Allow',
+                Principal: '*',
+                Resource: '*',
+              },
+            ],
+            Version: '2012-10-17',
+          },
+          {
+            Statement: [
+              {
+                Action: 'kms:action2',
+                Effect: 'Allow',
+                Principal: '*',
+                Resource: '*',
+              },
+            ],
+            Version: '2012-10-17',
+          },
+        ),
+      });
+    });
+
+    test('throws a descriptive exception', () => {
+      expect(() => {
+        kms.Key.fromCfnKey(cfnKey);
+      }).toThrow(/Could not parse the PolicyDocument of the passed AWS::KMS::Key/);
+    });
+  });
+
+  describe("called with a CfnKey that has an 'Fn::If' passed as the Statement of a KeyPolicy", () => {
```

Comment:
these tests may be over-indexing on the behaviour of `PolicyDocument.fromJson`. I haven't seen much mocking of other CDK modules but being able to control the return value of that function to write these tests might be nice

Reply:
I literally couldn't disagree more 😃.

`PolicyDocument.fromJson()` is an implementation detail, and has no place in these tests.

These tests actually show the 4 basic cases of handling dynamic values inside the Policy. While all 4 throw today, I can definitely see merit in making the last 2 not throw - either in this PR, or perhaps later, if we get feedback from customers that the current restrictions are too limiting.

The first 2 will be very difficult to make not throw, though.